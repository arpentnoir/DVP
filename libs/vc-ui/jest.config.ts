/* eslint-disable */
export default {
  displayName: 'vc-ui',
  preset: '../../jest.preset.js',
  setupFiles: ['./jest.setup.ts'],
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  coverageDirectory: '../../coverage/libs/vc-ui',
  moduleNameMapper: {
    '^react-pdf': 'react-pdf/dist/umd/entry.jest',
  },
};
