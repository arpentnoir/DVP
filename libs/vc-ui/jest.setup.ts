import 'jest-canvas-mock';
import '@testing-library/jest-dom';

const mockClipboard = {
  writeText: jest.fn(),
};

(global.navigator.clipboard as any) = mockClipboard;
