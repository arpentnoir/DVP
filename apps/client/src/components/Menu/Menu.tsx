import { Button, Text } from '@dvp/vc-ui';
import { Menu as MenuIcon } from '@mui/icons-material';
import CloseIcon from '@mui/icons-material/Close';
import NavigateBeforeIcon from '@mui/icons-material/NavigateBefore';
import {
  Box,
  ButtonBase,
  Divider,
  Drawer,
  Stack,
  useTheme,
} from '@mui/material';
import { Container } from '@mui/system';
import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';

export const MenuBar = React.forwardRef<HTMLButtonElement>((_props, ref) => {
  const [open, setOpen] = useState(false);

  const theme = useTheme();

  return (
    <React.Fragment>
      <Stack
        direction="row"
        sx={{
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor: theme.palette.common.backgroundBlue,
          height: '100%',
        }}
      >
        <Button
          data-testid="menu-button"
          ref={ref}
          sx={{
            height: '100%',
            padding: 2,
            color: 'white',
            boxShadow: 'none',
            backgroundColor: theme.palette.common.backgroundBlue,
            ':hover': {
              boxShadow: 'none',
              backgroundColor: theme.palette.common.tabBlueDark,
            },
          }}
          onClick={() => setOpen(true)}
          aria-expanded={open}
          aria-label="Menu"
          label="MENU"
          leftIcon={
            <MenuIcon
              sx={{
                fontSize: { xs: 30, sm: 40 },
              }}
            />
          }
          textProps={{
            sx: {
              display: { xs: 'none', sm: 'block' },
              fontSize: theme.typography.button.fontSize,
              fontFamily: theme.typography.button.fontFamily,
            },

            style: {
              marginLeft: '8px',
            },
          }}
        ></Button>
      </Stack>
      <MenuSide open={open} handleClose={() => setOpen(false)} />
    </React.Fragment>
  );
});

export const MenuSide = ({
  open,
  handleClose,
}: {
  open: boolean;
  handleClose: () => void;
}) => {
  const navigate = useNavigate();
  const theme = useTheme();

  const navigateTo = (path: string) => () => {
    handleClose();
    navigate(path);
  };

  return (
    <Drawer
      anchor="left"
      open={open}
      onClose={handleClose}
      PaperProps={{
        sx: {
          height: 'fit-content',
        },
      }}
    >
      <Container
        sx={{ width: { xs: '100vw', sm: '400px' } }}
        disableGutters
        aria-label="Side Menu"
      >
        <Stack direction="row">
          <ButtonBase
            sx={{
              color: theme.palette.common.backgroundDark,
              backgroundColor: theme.palette.common.featureGold,
              padding: '20px 24px',
              ':hover': {
                backgroundColor: theme.palette.common.featureGoldDark,
              },
            }}
            onClick={handleClose}
            tabIndex={1}
            aria-label="Close Menu"
          >
            <CloseIcon fontSize="large" />
            <Text
              paddingLeft="6px"
              fontSize={theme.typography.button.fontSize}
              fontFamily={theme.typography.button.fontFamily}
            >
              MENU
            </Text>
          </ButtonBase>
          <ButtonBase
            sx={{
              color: theme.palette.common.backgroundDark,
              backgroundColor: theme.palette.common.neutralGrey,
              padding: '20px 24px',
              flex: 1,
              ':hover': {
                backgroundColor: 'white',
              },
            }}
            onClick={() => window.open('https://homeaffairs.gov.au', '_blank')}
            tabIndex={1}
            aria-label="Close Menu"
          >
            <NavigateBeforeIcon fontSize="large" />
            <Text
              lineHeight="22px"
              paddingLeft="6px"
              textAlign="left"
              fontSize={theme.typography.button.fontSize}
              fontFamily={theme.typography.button.fontFamily}
            >
              HOME AFFAIRS PORTFOLIO
            </Text>
          </ButtonBase>
        </Stack>
        <Stack alignItems="flex-start">
          <MenuBarItem label={'Home'} handleClick={navigateTo('/')} />
          <MenuBarItem label={'Help'} handleClick={navigateTo('/')} />
          <MenuBarItem label={'Verify'} handleClick={navigateTo('/verify')} />
          <MenuBarItem
            label={'Documents'}
            handleClick={navigateTo('/documents')}
          />
          <MenuBarItem
            label={'Identities'}
            handleClick={navigateTo('/identities')}
          />
          <MenuBarItem label={'Issuers'} handleClick={navigateTo('/issuers')} />
          <MenuBarItem
            label={'Document schemas'}
            handleClick={navigateTo('/schemas')}
            showDivider={false}
          />
        </Stack>
      </Container>
    </Drawer>
  );
};

export const MenuBarItem = ({
  label,
  handleClick,
  showDivider = true,
}: {
  label: string;
  handleClick: () => void;
  showDivider?: boolean;
}) => {
  const theme = useTheme();

  return (
    <Box sx={{ width: '100%' }}>
      <ButtonBase
        data-testid={`menu-item:${label}`}
        sx={{
          color: theme.palette.common.backgroundBlue,
          '&:hover': {
            background: theme.palette.common.backgroundBlue,
            color: 'white',
            fontWeight: '600',
          },
          fontWeight: '500',
          fontSize: '20px',
          lineHeight: '30px',
          padding: '10px 26px',
          justifyContent: 'space-between',
          width: '100%',
          paddingX: '20px',
          height: '54px',
        }}
        onClick={handleClick}
        tabIndex={1}
      >
        <Text fontWeight="inherit" variant="body1">
          {label}
        </Text>
      </ButtonBase>
      {showDivider && <Divider />}
    </Box>
  );
};
