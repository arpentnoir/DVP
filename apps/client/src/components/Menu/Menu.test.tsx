import { render } from '@dvp/vc-ui';
import { fireEvent, waitFor } from '@testing-library/react';
import { MenuBar } from './Menu';

const mockNavigate = jest.fn();

// eslint-disable-next-line @typescript-eslint/no-unsafe-return
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

const menuItemNames = {
  Home: '/',
  Help: '/',
  Verify: '/verify',
  Documents: '/documents',
};

describe('Menu', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<MenuBar />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should show list of links when open', async () => {
    const { getByTestId, queryByTestId } = render(<MenuBar />);

    Object.keys(menuItemNames).forEach((menuItem) => {
      expect(queryByTestId(`menu-item:${menuItem}`)).toBeFalsy();
    });

    fireEvent.click(getByTestId('menu-button'));

    await waitFor(() =>
      Object.keys(menuItemNames).forEach((menuItem) => {
        expect(queryByTestId(`menu-item:${menuItem}`)).toBeTruthy();
      })
    );
  });

  it('should navigate to correct route for each link', () => {
    const { getByTestId } = render(<MenuBar />);

    fireEvent.click(getByTestId('menu-button'));

    Object.entries(menuItemNames).forEach(([name, path]) => {
      fireEvent.click(getByTestId(`menu-item:${name}`));

      expect(mockNavigate).toHaveBeenCalledWith(path);
    });
  });
});
