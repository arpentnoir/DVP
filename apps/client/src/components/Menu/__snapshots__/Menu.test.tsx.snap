// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render correctly 1`] = `
<body>
  <div>
    <div
      class="css-phqqqs-MuiStack-root"
    >
      <button
        aria-expanded="false"
        aria-label="Menu"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1nm1vzb-MuiButtonBase-root-MuiButton-root"
        data-testid="menu-button"
        tabindex="0"
        type="button"
      >
        <div
          class="css-doi3ij-MuiStack-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-4fyam7-MuiSvgIcon-root"
            data-testid="MenuIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body2 css-1dnu2q9-MuiTypography-root"
            data-testid="button:Menu:label"
            style="margin-left: 8px;"
            tabindex="0"
          >
            Menu
          </p>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;
